doc/developer/temporal-filters.md

```diff
@@ -0,0 +1,67 @@
+Understanding what Materialize does do may help you understand the behavior of and restrictions on temporal filters.
+
+## Manipulating timestamped updates
+
+Materialize records *changes* to collections of data.
+Each change occurs at some "logical timestamp", an explicit transcription of the time at which a change should occur.
+Generally, an update occurs at its logical timestamp and is then in effect indefinitely.
+To undo an update one issues a new update with a negated sign, at a future logical timestamp.
+
+We can restrict the duration of an update, to any non-empty interval `[from, until]`, by
+1. advancing the update's timestamp to at least `from`, and
+2. introducing a new negated update at timestamp `until`.
+The first step ensures that the update does not happen until `from` and the second step ensures that the update is retracted at `until`.
+Each of the two steps is optional, if either of the `from` or `until` bounds are unspecified.
+
+This mechanism allows us to support `WHERE` clauses in which `mz_logical_timestamp()` is directly related to an expression not containing `mz_logical_timestamp()`, or conjunctions (`AND`) of these clauses.
+The following query demonstrates several example uses:
+```sql
+SELECT *
+FROM data
+WHERE mz_logical_timestamp() BETWEEN data.valid_from AND data.valid_until
+  AND mz_logical_timestamp() > 100
+  AND 567 > mz_logical_timestamp();
+```
+The `BETWEEN` expression simplifies to two inequalities, and the other two inequalities are one-sided bounds.
+The `mz_logical_timestamp()` call can appear on either side of the inequality, but it must occur by itself.
+The inequality cannot be "not equal" (`!=` or `<>`) as this results in a "hole" rather than an upper and lower bound.
+
+## Correctness
+
+Using a temporal filter does not ensure that a record will be present at `valid_from`.
+The update that introduces a record has its own logical timestamp, and a temporal filter can only advance that time.
+To ensure precise validity consider using a CDC format that allows you to specify update timestamps explicitly.
+
+## Performance
+
+Materialize will re-evaluate downstream views at each timestamp presented to them.
+These timestamps are denominated in milliseconds, but are often much more coarse-grained because of a source's timestamping policy.
+For example, the default source timestamping frequency is once every second.
+This means that downstream views will only see the collection change once per second, and e.g. each aggregate value will change at most once per second.
+This reduces the computational burden over a system whose values changed every millisecond.
+
+Temporal filters allow you to introduce updates with timestamps that do not otherwise exist in the input.
+If your `valid_from` and `valid_until` columns are arbitrary milliseconds not aligned to seconds, you may prompt many additional recomputations and substantially impair Materialize's performance.
```

Comment:
is it worth it to round the from / until bounds to the nearest lower / upper multiple of timestamp frequency?

Reply:
Not automatically, imo, but that might be a good idea to recommend in this doc.